enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,328 @@
+
+  `KernelURL` to store the matching kernel.
+
+  `ExtraKernelParams` to provide a link to the root file system. BMO will be
+  updated to pass the parameters to the node.
+
+- [Ironic agent][ironic-agent] currently contains a custom deployment plugin
+  (*custom deploy method* in baremetal-operator terms, *custom deploy step* in
+  Ironic terms) that invokes `coreos-installer` for the image installation.
+  A new custom deploy method `start_assisted_install` will be added to start
+  the assisted agent when BMO requests deployment.
+
+- The assisted service (more specifically, BMAC) will stop explicitly disabling
+  *inspection* on `BareMetalHost`s it manages, will start using a different
+  process based on the new `start_assisted_install` deploy method and stop
+  rebooting the machine in the end of the process.
+
+### User Stories
+
+This change itself does not add new user stories, but rather enables existing
+features to work for ZTP:
+
+- As an operator, I want to ensure that all OpenShift nodes are deployed with
+  known firmware settings.
+
+- As an operator of hardware that does not support virtual media, I want to
+  deploy OpenShift nodes using iPXE network boot.
+
+- As an operator, I want OpenShift to ensure that the deployment ISO is
+  correctly disconnected on installation finish and cannot be accidentally
+  booted into.
+
+### API Extensions
+
+The `PreprovisioningImage`'s `Status` will be updated with two new fields:
+
+- `KernelURL` - the URL of a kernel image to use with (i)PXE boot.
+
+- `ExtraKernelParams` - additional kernel parameters to pass when booting
+  using (i)PXE.
+
+### Implementation Details/Notes/Constraints
+
+#### Coordination between agents
+
+The first agent to start will be the Ironic agent since it is responsible
+for coordinating the whole flow and (indirectly through Ironic) updating the
+`BareMetalHost`'s status. *Inspection* and additional configuration (RAID etc)
+will be handled by the Ironic agent alone.
+
+BMAC will be able to stop synchronizing the inventory to `BareMetalHost`
+objects since this task will be handled by Ironic and baremetal-operator.
+
+Currently, the provisioning is initiated by BMAC (the assisted component
+responsible for interactions with baremetal-operator) by applying the following
+change to the `BareMetalHost`'s spec:
+
+```yaml
+Image:
+  URL: url/of/assisted/agent.iso
+  DiskFormat: live-iso
+```
+
+After this enhancement, BMAC will instead apply:
+
+```yaml
+CustomDeploy:
+  Method: start_assisted_install
+```
+
+This instructs Ironic to find and execute a *custom deploy step* called
+`start_assisted_install`, which will be provided by an OpenShift specific
+Ironic agent plugin. This step will use D-Bus to order systemd to start the
+assisted agent and watch its status.
+
+If the assisted agent exits successfully, the Ironic agent will report success
+to Ironic and baremetal-operator. Otherwise, a failure will be recorded.
```

Comment:
what will be done with failure? I mean what status we will have in the host? how user should continue from that state? Just not sure what error mean in that case

Reply:
It's going to propagate to the BMH object (which will probably cause BMO to retry, but I don't want this document to list all the details about BMO).